swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -558,6 +587,29 @@ predicate storeStep(Node node1, ContentSet c, Node node2) {
     c.isSingleton(any(Content::TupleContent tc | tc.getIndex() = tuple.getIndex()))
   )
   or
+  // creation of an enum `.variant(v1, v2)`
+  exists(EnumElementExpr enum, int pos |
+    node1.asExpr() = enum.getArgument(pos).getExpr() and
+    node2.asExpr() = enum and
+    c.isSingleton(any(Content::EnumContent ec | ec.getParam() = enum.getElement().getParam(pos)))
+  )
+  or
+  // creation of an optional via implicit conversion,
+  // i.e. from `f(x)` where `x: T` into `f(.some(x))` where the context `f` expects a `T?`.
+  exists(InjectIntoOptionalExpr e |
+    e.convertsFrom(node1.asExpr()) and
+    node2 = node1 and // HACK: we should ideally have a separate Node case for the (hidden) InjectIntoOptionalExpr
+    c instanceof OptionalSomeContentSet
+  )
+  or
+  // creation of an optional by returning from an optional initializer (`init?`)
+  exists(ConstructorDecl init, OptionalType initRetType |
+    node1.asExpr().(CallExpr).getStaticTarget() = init and
+    node2 = node1 and // HACK: again, we should ideally have a separate Node case here, and not reuse the CallExpr
+    c instanceof OptionalSomeContentSet and
+    init.getInterfaceType().(FunctionType).getResult().(FunctionType).getResult() = initRetType
```

Comment:
Is this just a really complicated way of saying `init` is an `init?` constructor? If so, could we move that to a different class? Maybe even an `AST` class?

Reply:
Moved it into ConstructorDecl.isFailable (the official Swift terminology)